lucene/core/src/java/org/apache/lucene/codecs/lucene80/IndexedDISI.java

```diff
@@ -0,0 +1,542 @@
+   * @param jumpTableEntryCount the number of blocks convered by the jump-table.
+   * @param cost normally the number of logical docIDs.
+   */
+  IndexedDISI(IndexInput in, long offset, long length, int jumpTableEntryCount, long cost) throws IOException {
+    this(in.slice("docs", offset, length), jumpTableEntryCount, cost);
+  }
+
+  /**
+   * This constructor allows to pass the slice directly in case it helps reuse.
+   * see eg. Lucene80 norms producer's merge instance.
+   * @param slice backing data, starting from position 0 to slice.length
+   * @param jumpTableEntryCount the number of blocks convered by the jump-table.
+   * @param cost normally the number of logical docIDs.
+   */
+  IndexedDISI(IndexInput slice, int jumpTableEntryCount, long cost) throws IOException {
+    this.slice = slice;
+    this.cost = cost;
+
+    this.jumpTableEntryCount = jumpTableEntryCount;
+    //slice.seek(slice.length()-Short.BYTES);
+    if (jumpTableEntryCount <= 0) {
+      jumpTable = null;
+    } else {
+      long jumpTableOffset = slice.length()-jumpTableEntryCount*Long.BYTES;
+      jumpTable = slice.randomAccessSlice(jumpTableOffset, slice.length()-jumpTableOffset);
+    }
+  }
+
+  private int block = -1;
+  private long blockEnd;
+  private long denseBitmapOffset = -1; // Only used for DENSE blocks
+  private int nextBlockIndex = -1;
+  Method method;
+
+  private int doc = -1;
+  private int index = -1;
+
+  // SPARSE variables
+  private boolean exists;
+
+  // DENSE variables
+  private long word;
+  private int wordIndex = -1;
+  // number of one bits encountered so far, including those of `word`
+  private int numberOfOnes;
+  // Used with rank for jumps inside of DENSE as they are absolute instead of relative
+  private int denseOrigoIndex;
+
+  // ALL variables
+  private int gap;
+
+  @Override
+  public int docID() {
+    return doc;
+  }
+
+  @Override
+  public int advance(int target) throws IOException {
+    final int targetBlock = target & 0xFFFF0000;
+    if (block < targetBlock) {
+      advanceBlock(targetBlock);
+    }
+    if (block == targetBlock) {
+      if (method.advanceWithinBlock(this, target)) {
+        return doc;
+      }
+      readBlockHeader();
+    }
+    boolean found = method.advanceWithinBlock(this, block);
+    assert found;
+    return doc;
+  }
+
+  public boolean advanceExact(int target) throws IOException {
+    final int targetBlock = target & 0xFFFF0000;
+    if (block < targetBlock) {
+      advanceBlock(targetBlock);
+    }
+    boolean found = block == targetBlock && method.advanceExactWithinBlock(this, target);
+    this.doc = target;
+    return found;
+  }
+
+  private void advanceBlock(int targetBlock) throws IOException {
+    final int blockIndex = targetBlock >> 16;
+    // If the destination block is 2 blocks or more ahead, we use the jump-table.
+    if (jumpTable != null && blockIndex >= (block >> 16)+2 && blockIndex < jumpTableEntryCount) {
```

Comment:
This seems to mean that if you call `advance(target)` on a target that is greater than the higher docID of the last block, then you will resort to a linear scan. Can we fix it?

As a side note, `jumpTable != null` seems to be covered by `blockIndex < jumpTableEntryCount` given that jumpTable is only null if jumpTableEntryCount is zero?